packages/flutter/lib/src/rendering/object.dart

```diff
@@ -4188,6 +4187,62 @@ mixin RenderObjectWithChildMixin<ChildType extends RenderObject> on RenderObject
   }
 }
 
+/// A mixin for [RenderObject] subclasses with a layout callback. The mixin
+/// guarantees the layout callback will be called even if this [RenderObject]
+/// skips doing layout.
+///
+/// A layout callback is a callback that mutates the [RenderObject]'s render
+/// subtree, invoked within an [invokeLayoutCallback] during the [RenderObject]'s
+/// layout process. Sometimes such callbacks involve rebuilding dirty widgets,
+/// especially if the widget configuration depends on the layout process (most
+/// notably, the [LayoutBuilder] widget). Unlike render subtrees, typically all
+/// dirty widgets (even the off-screen ones) in a widget tree must be rebuilt.
+/// This mixin makes sure the layout callback will be invoked (such that dirty
+/// [Element]s are rebuilt) even when an ancestor [RenderObject] chose to skip
+/// laying out this [RenderObject].
+///
+/// Subclasses must not invoke the layout callback directly. Instead, call
+/// `super.performLayout` in the [performLayout] implementation.
+///
+/// See also:
+///
+///  * [LayoutBuilder] and [SliverLayoutBuilder], which use the mixin.
+mixin RenderObjectWithLayoutCallbackMixin on RenderObject {
+  // The initial value of this flag must be set to true to prevent the layout
+  // callback from being called when the tree has never been laid out.
+  bool _needsRebuild = true;
+
+  @mustCallSuper
+  @override
+  void performLayout() {
+    invokeLayoutCallback(runLayoutCallback);
+    _needsRebuild = false;
+  }
+
+  /// The layout callback to be invoked during [performLayout].
+  ///
+  /// This method should not be invoked directly. Instead, call
+  /// `super.performLayout` in the [performLayout] implementation. This
+  /// implementation will be invoked within [invokeLayoutCallback].
+  @visibleForOverriding
+  void runLayoutCallback(covariant Constraints constraints);
+
+  /// Informs the framework that the layout callback has been updated and must
+  /// be invoked again.
+  void scheduleLayoutCallback() {
+    if (_needsRebuild) {
+      assert(debugNeedsLayout);
+      return;
+    }
+    _needsRebuild = true;
+    // This ensures that the layout callback will be run even if an ancestor
+    // chooses to not lay out this subtree (for example, OverlayEntries with
+    // `maintainState` set to true).
+    owner?._nodesNeedingLayout.add(this);
```

Comment:
For my own understanding: Wouldn't this case child nodes to potentially also get layed out unnecessary?

Reply:
Or is the idea that since the contrains haven't changed, the layout builder is pretty much guaranteed to pass the same constraints to its child and layout would short-circuit right there?

Reply:
Yeah the render children may also do layout now which can be unnecessary. But I think (at least for `RenderTheater`) if you rebuild a widget subtree in an inactive `maintainState` OverlayEntry, relayout boundaries in the render subtree are still operational and call `performLayout` as usual?

Reply:
Is the reason that we don't want to find the closest boundary because we are not sure whether between that and here may skips the layout?

Is it possible to short circuit the layout pass and directly marksNeedsBuild of the subtree if this is in active route? I am a little concern about removing the check that layout can start at non layout boundary.

Reply:
> Is it possible to short circuit the layout pass and directly marksNeedsBuild of the subtree if this is in active route? I am a little concern about removing the check that layout can start at non layout boundary.

I am not aware of any other ways to detect whether a `LayoutBuilder`s layout has been skipped, other than adding its node to the dirtly layout list.

If the parent skips layout, then it doesn't care about the size of this render object so this render object is effectively a relayout boundary.

Reply:
yeah, I can't think of a better way to do this. especially we only want to rebuild during layout phase .